design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+-- the query conceptually "scans" the full inputTable (in particular the current generation when the query is issued),
+   computes the finite result, and terminates
+```
+
+
+**Example 250: a table-table join**
+
+```sql
+[CREATE TABLE myRdbmsStyleTable AS] SELECT * FROM inputTable t1 JOIN otherTable t2 ON t1.key = t2.key
+-- returns a TABLE
+-- same properties as Example 200
+-- because all inputs are a table, the output is a table
+```
+
+The last point is a design principle (cf. P2 above)
+
+* (P3) a query with only TABLES as inputs, returns a TABLE as output
+* (P3-a) we can extend this principle to n-ary TABLE joins
+
+An open question is, what table version/generations are joined to each other for this case – for unary queries,
+we can just use the "current generation of the current version" (i.e., whatever is in RocksDB atm when the query is issued).
+For n-ary queries, we need to figure out what semantics we want to have/allow (if we ever support this type of join to begin
+with—--currently, I would advocate to not allow this query at all, but tell users to define a continuous table-table join query with proper
+event-time semantics and query the result TABLE/MV instead).
+
+
+**Example 300: creating a materialized view (persistent continuous query of an input table)**
+
+```sql
+CREATE MATERIALIZED VIEW resultView AS SELECT * FROM inputTable
+-- returns a TABLE
+-- this is always a persistent query (without the CREATE MV AS prefix, the query is the same as Example 200 – compare Example 400 for the transient version)
+-- the result is a MV, ie, a table with a continuous queries that update the table when the base data changes; insert/update/delete statements are not allowed
+-- the same as current CTAS; for backward compatibility, current CTAS would still be supported but would be deprecated
+```
+
+
+**Example 400: subscribing to a changelog stream (transient continuous query of an input table)**
+
+```sql
+[CREATE STREAM myChangelog AS] SELECT * FROM STREAM(tableOrMaterializedView)
+-- returns a STREAM
+-- compare Example 100 (those properties apply)
+-- without CSAS prefix, it is the transient query counterpart to Example 300 – all updates are "streamed" continuously to the client
+--   (note, that the result is a fact-stream and it's the client responsibility to "interpret" the records with update semantics)
+-- The STREAM(...) operator creates a changelog stream from its input TABLE (can be a RDBMS style table or a MV); compare Example 0
+-- The output stream contains one record for each record in the table (ie, full table scan) based on the table generation when the query was issued
+--   plus a record for all future updates
+```
+
+
+**Example 500: aggregating a stream**
+
+```sql
+[CREATE MATERIALIZED VIEW resultView AS] SELECT count(*) FROM inputStream GROUP BY <some-grouping-condition>
+-- returns a STREAM (!!!): different and not different to current KSQL semantics... compare below
```

Comment:
> To perform an aggregation it is necessary to materialize state in some form of table. How can I access that table? 

You can't because is a `TEMP TABLE` -- RDBMS systems have the same concept.

> If I need to re-declare it over the changelog topic we need to have a really smart optimizer to know to leverage the existing one

Yes, but this seems to be standard query optimization.

> and this metadata might be lost when I move between KSQL clusters.

Cannot follow here? Can you elaborate?

Reply:
I agree, it's a temp table, (or hidden state store). 

> If I need to re-declare it over the changelog topic we need to have a really smart optimizer to know to leverage the existing one

Do you mean if someone executed the same statement again, e.g: issuing:

``` sql
[CREATE MATERIALIZED VIEW resultView AS] SELECT count(*) FROM inputStream GROUP BY <some-grouping-condition>
```
 multiple times? (with different view name)

If its a transient query then each instance would need its own temp table, as each would be streaming back the changes as the aggregate count increased.

If it's an MV, then initially we'd probably create a duplicate copy, but it wouldn't be two hard to detect if the query is _exactly_ the same and to make the second view an alias of the first, (with reference counting obvs).

Or do you mean something else?

>  and this metadata might be lost when I move between KSQL clusters.

Yep, I'm lost here too...